### ISSUE TITLE
numpy.savetxt cannot save file correctly using file handle 

### ISSUE THREAD

**Issue by shelper** [numpy/numpy#10018]
with the newest numpy 1.13.3, if I have an array called points, with shape m x 6 and dtype of float32. I can save the array to a "foo.txt" file as below:

```
np.savetxt('foo.txt', points, fmt='%f %f %f %d %d %d')
```
but if I run
```
with open('foo.txt', 'w') as f:
    np.savetxt(f, points, fmt='%f %f %f %d %d %d')
```
i got the error below:

> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> ~\AppData\Local\Continuum\Anaconda3\lib\site-packages\numpy\lib\npyio.py in savetxt(fname, X, fmt, delimiter, newline, header, footer, comments)
>    1214                 try:
> -> 1215                     fh.write(asbytes(format % tuple(row) + newline))
>    1216                 except TypeError:
> 
> TypeError: write() argument must be str, not bytes
> 
> During handling of the above exception, another exception occurred:
> 
> TypeError                                 Traceback (most recent call last)
> <ipython-input-280-ffd2ddd0719a> in <module>()
>       1 with open('foo.txt', 'w') as f:
> ----> 2     np.savetxt(f, points, fmt='%f %f %f %d %d %d')
>       3 
> 
> ~\AppData\Local\Continuum\Anaconda3\lib\site-packages\numpy\lib\npyio.py in savetxt(fname, X, fmt, delimiter, newline, header, footer, comments)
>    1217                     raise TypeError("Mismatch between array dtype ('%s') and "
>    1218                                     "format specifier ('%s')"
> -> 1219                                     % (str(X.dtype), format))
>    1220         if len(footer) > 0:
>    1221             footer = footer.replace('\n', '\n' + comments)
> 
> TypeError: Mismatch between array dtype ('float32') and format specifier ('%f %f %f %d %d %d')

**Comment by pdebuyl** [comment]
The already open file handle should be binary here. Try:

``` .py
with open('foo.txt', 'wb') as f:
    np.savetxt(f, points, fmt='%f %f %f %d %d %d')
```

Note the behaviour of file handles might differ between Python 2 and Python 3. Other note: your format specifier '%d' is for integers. Can you let us know if the problem is resolved with the fix above?

**Comment by shelper** [comment]
thanks it is resolved,

**Comment by stefan-c-kremer** [comment]
Wait!  np.savetxt will not save to a a text file (it has to be binary?).  This seems completely wrong!  I am writing to a text file that I opened in "w" mode.  I write some strings to it.  Now I want to write a np vector to it in text format, but I can't because savetxt doesn't write text.  But I also can't open the file in binary mode because then I can't write strings to it.  Shouldn't savetxt work on txt files?

**Comment by maximilianh** [comment]
I agree with stefan-c-kremer. What is worse is that - as pdebuyl said - there doesn't seem to be a way to make a piece of code work on both python2 and python3. If I open it with "wb", I get an error with python3. If I open it with "w", I get an error with python2.

**Comment by mwtoews** [comment]
For the record, this issue was resolved with numpy version 1.14.0.

**Comment by stefan-c-kremer** [comment]
Thank you.  Looking forward to trying it out.
